File: algorithms/dijkstra/go/dijkstra.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

type Edge struct {
	To     int
	Weight int
}

type heapEntry struct {
	node int
	dist int
}

type minHeap []heapEntry

func (h minHeap) Len() int            { return len(h) }
func (h minHeap) Less(i, j int) bool  { return h[i].dist < h[j].dist }
func (h minHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *minHeap) Push(x any) { *h = append(*h, x.(heapEntry)) }
func (h *minHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

// グラフ定義に現れる全ノード（到達不可能なものを含む）を収集する。
// 戻り値の dist には未到達ノードも math.MaxInt で含まれる仕様。
func collectNodes(graph map[int][]Edge, start int) map[int]bool {
	nodes := map[int]bool{start: true}
	for node, edges := range graph {
		nodes[node] = true
		for _, e := range edges {
			nodes[e.To] = true
		}
	}
	return nodes
}

func Dijkstra(graph map[int][]Edge, start int) map[int]int {
	for _, edges := range graph {
		for _, e := range edges {
			if e.Weight < 0 {
				panic(fmt.Errorf("negative edge weight %d is not supported", e.Weight))
			}
		}
	}
	nodes := collectNodes(graph, start)
	dist := make(map[int]int, len(nodes))
	for node := range nodes {
		dist[node] = math.MaxInt
	}
	dist[start] = 0

	pq := &minHeap{{node: start, dist: 0}}
	heap.Init(pq)

	for pq.Len() > 0 {
		entry := heap.Pop(pq).(heapEntry)
		u, du := entry.node, entry.dist
		if du > dist[u] {
			continue
		}
		for _, e := range graph[u] {
			if du > math.MaxInt-e.Weight {
				continue
			}
			newDist := du + e.Weight
			if newDist < dist[e.To] {
				dist[e.To] = newDist
				heap.Push(pq, heapEntry{node: e.To, dist: newDist})
			}
		}
	}

	return dist
}

File: algorithms/dijkstra/typescript/dijkstra.ts
export type Edge = { to: number; weight: number }
export type WeightedGraph = Map<number, Edge[]>

type HeapEntry = { node: number; dist: number }

// グラフ定義に現れる全ノード（到達不可能なものを含む）を収集する。
// 戻り値の dist には未到達ノードも Infinity で含まれる仕様。
function collectNodes(graph: WeightedGraph, start: number): Set<number> {
  const nodes = new Set<number>([start])
  for (const [node, edges] of graph) {
    nodes.add(node)
    for (const { to } of edges) nodes.add(to)
  }
  return nodes
}

class MinHeap {
  private data: HeapEntry[] = []

  get size() {
    return this.data.length
  }

  push(entry: HeapEntry): void {
    this.data.push(entry)
    this.bubbleUp(this.data.length - 1)
  }

  pop(): HeapEntry | undefined {
    if (this.data.length === 0) return undefined
    const top = this.data[0]
    const last = this.data.pop()!
    if (this.data.length > 0) {
      this.data[0] = last
      this.sinkDown(0)
    }
    return top
  }

  private bubbleUp(i: number): void {
    while (i > 0) {
      const parent = (i - 1) >> 1
      if (this.data[parent].dist <= this.data[i].dist) break
      ;[this.data[parent], this.data[i]] = [this.data[i], this.data[parent]]
      i = parent
    }
  }

  private sinkDown(i: number): void {
    const n = this.data.length
    while (true) {
      let smallest = i
      const left = 2 * i + 1
      const right = 2 * i + 2
      if (left < n && this.data[left].dist < this.data[smallest].dist) smallest = left
      if (right < n && this.data[right].dist < this.data[smallest].dist) smallest = right
      if (smallest === i) break
      ;[this.data[smallest], this.data[i]] = [this.data[i], this.data[smallest]]
      i = smallest
    }
  }
}

export function dijkstra(graph: WeightedGraph, start: number): Map<number, number> {
  for (const edges of graph.values()) {
    for (const { weight } of edges) {
      if (weight < 0) throw new RangeError(`negative edge weight ${weight} is not supported`)
    }
  }
  const nodes = collectNodes(graph, start)
  const dist = new Map<number, number>()
  for (const node of nodes) dist.set(node, Infinity)
  dist.set(start, 0)

  const heap = new MinHeap()
  heap.push({ node: start, dist: 0 })

  while (heap.size > 0) {
    const { node: u, dist: du } = heap.pop()!
    if (du > dist.get(u)!) continue

    for (const { to, weight } of graph.get(u) ?? []) {
      const newDist = du + weight
      const toDist = dist.get(to)!
      if (newDist < toDist) {
        dist.set(to, newDist)
        heap.push({ node: to, dist: newDist })
      }
    }
  }

  return dist
}
